text/3806-align-attr.md

```diff
@@ -0,0 +1,570 @@
+
+The layout of a `repr(C_for_real)` ADT with `align` attributes on its fields is
+identical to that of the corresponding C ADT declared with `alignas`
+annotations. For example, the struct below is equivalent to the C `struct foo`
+from the motivation section:
+
+```rust
+#[repr(C_for_real)]
+pub struct foo {
+    pub x: u8,
+    #[align(128)]
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+### `repr(linear)`
+
+In a `repr(linear)` ADT, a field with an `align` attribute has its alignment, as
+well as the alignment of the containing ADT, increased to at least what the
+attribute specifies.
+
+For example, the following two structs have the same layout in memory (though
+not necessarily the same ABI):
+
+```rust
+#[repr(linear)]
+pub struct foo {
+    pub x: u8,
+    #[align(128)]
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+```rust
+#[repr(linear, align(128))]
+pub struct foo2 {
+    pub x: u8,
+    pub _padding: [MaybeUninit<u8>; 127usize],
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+## On `static`s
+
+Any `static` item (including `static`s inside `extern` blocks) may have an
+`align` attribute applied:
+
+```rust
+#[align(32)]
+static BAZ: [u32; 12] = [0xDEADBEEF; 12];
+
+unsafe extern "C" {
+    #[align(2)]
+    safe static BOZZLE: u8;
+}
+```
+
+The effect of the attribute is to force the `static` to be stored with at least
+the specified alignment. The attribute does not force padding bytes to be added
+after the `static`. For `static`s inside `unsafe extern` blocks, if the `static`
+does not meet the specified alignment, the behavior is undefined. (This UB is
+analogous to the UB that can result if the static item is not a valid value of
+its type. The question of whether the UB can occur even if the item is unused,
+has the same answer for both cases.)
+
+The `align` attribute may also be applied to thread-local `static`s created with
+the `thread_local!` macro; the attribute affects the alignment of the underlying
+value, not that of the outer `std::thread::LocalKey`.
+
+```rust
+thread_local! {
+    #[align(64)]
+    static FOO: u8 = 42;
+}
+
+fn main() {
+    FOO.with(|r| {
+        let p: *const u8 = r;
+        assert_eq!(p.align_offset(64), 0);
+    });
+}
+```
+
+`align` attributes on `static`s are shown in `rustdoc`-generated documentation.
+
+## On function items
+
+On function items, `#[align(…)]` sets the alignment of the function’s code. (It
+does not affect the alignment of its function item type, which remains a 1-ZST.)
+This replaces `#[repr(align(…))]` on function items, from
+`#![feature(fn_align)]`.
+
+On `async fn`, the attribute controls the alignment of the code of the function
+that returns the `Future`.
+
+On function items in trait declarations, `#[align(…)]` specifies the minimum
+alignment that all implementations of the item must have. `impl` blocks
+containing the item must specify an `#[align(…)]` at least as high. The `dyn`
+implementation gemerated by the compiler must also provide this alignment. Any
```

Comment:
It kinda seems to me this constrains vtable layout without many clear benefits, and with perhaps surprising results? Multiple instances of the same vtable can be present in the same program and people write code that tries to pretend these are unique, but they aren't.

Reply:
Resolving this since in other messages I already asked "what's the use case?" in actually brainful ways :face_with_spiral_eyes: